### PR TITLE
region_cache: invalidate region after meeting store_not_match

### PR DESCRIPTION
**Pull request by 5kbpers**
Signed-off-by: 5kbpers <tangminghua@pingcap.com>

### What problem does this PR solve?

Problem Summary:

When a store goes offline and starting another store with the same address, some requests will retry infinitely after meeting the `store_not_match` error.

### What is changed and how it works?

Do not retry after meeting the `store_not_match` error.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- region_cache: fix infinitely retrying after meeting the `store_not_match` error

### CODE REVIEW

**Comment by 5kbpers** [comment]
@lysu @youjiali1995 @MyonKeminta PTAL

**Review by youjiali1995 — COMMENTED** [review]
> When a store goes offline and starting another store with the same address, ...

The store is stopped but isn't tombstone?

**Review by MyonKeminta — APPROVED** [review]
LGTM

**Comment by 5kbpers** [comment]
> > When a store goes offline and starting another store with the same address, ...
> 
> The store is stopped but isn't tombstone?

No, the store got offline and was added again.

File: store/tikv/region_request.go [hunk 1]
```diff
@@ -668,11 +668,11 @@ func (s *RegionRequestSender) onRegionError(bo *Backoffer, ctx *RPCContext, seed
 
 	if storeNotMatch := regionErr.GetStoreNotMatch(); storeNotMatch != nil {
 		// store not match
-		logutil.BgLogger().Debug("tikv reports `StoreNotMatch` retry later",
+		logutil.BgLogger().Debug("tikv reports `StoreNotMatch` return error",
 			zap.Stringer("storeNotMatch", storeNotMatch),
 			zap.Stringer("ctx", ctx))
 		ctx.Store.markNeedCheck(s.regionCache.notifyCheckCh)
-		return true, nil
+		return false, errors.New(regionErr.String())
```
Comment (youjiali1995) [inline store/tikv/region_request.go:675]: This error is reported to the user directly which is unacceptable.

Comment (youjiali1995) [inline store/tikv/region_request.go:675]:
```suggestion
		s.regionCache.InvalidateCachedRegion(ctx.Region)
		return true, nil
```

**Comment by youjiali1995** [comment]
https://github.com/pingcap/tidb/issues/22907 handles such a case, but TiDB doesn't use the PD client correctly. https://github.com/tikv/pd/blob/bc63de897afa69cfcc5828f45431fe3999fed4b3/client/client.go#L1159-L1168:
```go
func handleStoreResponse(resp *pdpb.GetStoreResponse) (*metapb.Store, error) {
	store := resp.GetStore()
	if store == nil {
		return nil, errors.New("[pd] store field in rpc response not set")
	}
	if store.GetState() == metapb.StoreState_Tombstone {
		return nil, nil
	}
	return store, nil
}
```

**Review by youjiali1995 — APPROVED** [review]

**Pull request merged**